### ISSUE TITLE
20210722: build fails on systems without iproute2 installed

### ISSUE THREAD

**Issue by arekm** [iputils/iputils#383]
build fails on systems without iproute2 installed and iproute seem to be only needed for test (so that ip shouldn't be needed just for building)

````
[...]
Run-time dependency threads found: YES
Has header "error.h" : YES
Program /usr/bin/meson found: YES (/usr/bin/meson)
Configuring config.h using configuration
Program /sbin/setcap found: YES (/sbin/setcap)

ping/meson.build:38:0: ERROR: Program or command 'ip' not found or not executable

A full log can be found at /tmp/B.x7dvybsu/BUILD/iputils-20210722/build/meson-logs/meson-log.txt
````

**Comment by pevik** [comment]
You can disable tests with `-DSKIP_TESTS=true`.

**Comment by arekm** [comment]
Doesn't work. Still wants ip binary.

**Comment by Berzerkula** [comment]
On Thu, 02 Dec 2021 04:45:29 -0800
Arkadiusz Miśkiewicz ***@***.***> wrote:

> uild fails on systems without iproute2 installed and iproute

Why are you living in the past?

Official notice was given.

https://lists.debian.org/debian-devel/2009/03/msg00780.html

ifconfig from inetutils or net-tools is not the proper way these days.

Sincerely,

William Harrington

**Comment by arekm** [comment]
I'm living in future. net-tools/inetutils doesn't exist here, too.

It's a build container which essential deps only. iproute2 is only needed for tests, so this bug report is valid. Testing if ip exists should hide inside of tests condition.

**Comment by pevik** [comment]
This works for me even without `ip`:

`meson builddir -DSKIP_TESTS=true && make`

It's really rare not have `ip`, thus nobody noticed that. What prevent you install iproute2 on the container? You needed to install other things anyway (compiler is much bigger than iproute2, which is kind of essential package anyway).

`ip` is used for IPv6 detection. Well, we could replace it with other method to detect, i.e. `cat /sys/module/ipv6/parameters/disable` (although there is really no difference - if you don't have coreutils nor busybox you will miss`cat` and `cat` is needed for testing right? :)).

**Comment by arekm** [comment]
But how that works for you?

ping/meson.build runs
````
r = run_command('ip', '-6', 'a')
````

ip command unconditionally (it's not inside any tests condition) and detects that ending with initially reported error (meson 0.59.2).

Of course I can locally patch things and disable this but it didn't make sense, so I've reported it.

If "that's the way" then ok, I'll just patch locally.

**Comment by pevik** [comment]
> ip command unconditionally (it's not inside any tests condition) and detects that ending with initially reported error (meson 0.59.2).

Not sure if I understand what you mean. What's your suggestion to fix?

**Comment by nmeyerhans** [comment]
The problem here is that the nonexistence of `ip` is not intended to be a fatal error, but it is. We use it to test for the presence of IPv6 connectivity, and proceed to run IPv4-only tests if `ip` indicates that we lack such connectivity. If the `ip` binary is not present, then it would be better to interpret that as a lack of IPv6 addresses rather than to fail the test run altogether.

I think a better test, since we really only care about `::1` being available, is to check `/proc/net/ipv6_route` for a local route to that destination. An IPv6 localhost route will appear as a line starting with `00000000000000000000000000000001` in that file.

**Comment by eli-schwartz** [comment]
You try to run the `ip` binary to do a configure-time check. There are two ways this can fail:
- the binary can return a non-zero exit status
- meson cannot find the binary to even try running it

The former case is handled by `check: false` as a kwarg to the run_command() function. Note that this is the default, but future versions of meson will emit a warning if you don't specify either true or false, and future versions of meson may make it default to true (because it's not obvious that running a subprocess during configure ignores errors by default, and this is often not wanted).

The latter case can be handled by not letting `run_command()` internally invoke `find_program()` on the program argument. Instead, use

```meson
ip = find_program('ip', required: false)
if ip.found() and run_command(ip, '-6', 'a').stdout().contains('::1')
  message('IPv6 enabled')
  ipv6_dst = [ '::1' ]
  ipv6_switch = [ '-6' ]
else
  message('WARNING: IPv6 disabled')
endif
```

BTW: since you are using contains() I highly doubt you need to use .strip(), removing that part of the function chain avoids useless work and helps keep line lengths down.

**Comment by pevik** [comment]
@eli-schwartz Looks like an improvement, feel free to post PR with this.

> The problem here is that the nonexistence of `ip` is not intended to be a fatal error, but it is.

BTW while open to improvements I still don't see why this is a hard dependency: even with removed `ip` it's possible to build (tested on Debian stable):

```
# mv /usr/bin/ip /usr/bin/ip_
# ip
-bash: /usr/bin/ip: No such file or directory
# rm -rf builddir/; meson builddir -DSKIP_TESTS=true && make
The Meson build system
Version: 0.56.2
Source dir: /root/iputils.git
Build dir: /root/iputils.git/builddir
Build type: native build
Project name: iputils
Project version: 20211215
C compiler for the host machine: cc (gcc 10.2.1 "cc (Debian 10.2.1-6) 10.2.1 20210110")
C linker for the host machine: cc ld.bfd 2.35.2
Host machine cpu family: x86_64
Host machine cpu: x86_64
Found pkg-config: /usr/bin/pkg-config (0.29.2)
Run-time dependency systemd found: YES 247
Checking for function "__fpending" : YES 
Checking for function "getrandom" : YES 
Checking for function "nanosleep" : YES 
Checking if "struct icmp6_nodeinfo" compiles: NO 
Checking if "struct tm time;" compiles: YES 
Library m found: YES
Library resolv found: YES
Checking for function "clock_gettime" : YES 
Library disabler-appears-to-disable-executable-build found: NO
Library cap found: YES
Library intl found: NO
Run-time dependency libidn2 found: YES 2.3.0
Has header "error.h" : YES 
Found git repository at /root/iputils.git
Configuring config.h using configuration
Program setcap found: YES (/usr/sbin/setcap)
Program xsltproc found: YES (/usr/bin/xsltproc)
Message: 

APPLICATIONS (build)
man: true
HTML man: false
arping: true (capability or suid: false)
clockdiff: true (capability or suid: false)
ping: true (capability or suid: false)
tracepath: true

CONFIGURATION
Capatiblity (with libcap): true
IDN (with libidn2): true
I18N (with gettext): true
systemd: true
run tests: false

SYSTEM PATHS
prefix: /usr/local
bindir: /usr/local/bin
localedir: /usr/local/share/locale
sbindir: /usr/local/sbin
systemdunitdir: /lib/systemd/system

Build targets in project: 13

Found ninja-1.10.1 at /usr/bin/ninja
ninja -C builddir
ninja: Entering directory `builddir'
[19/19] Linking target ping/ping

$ grep ipv6.disable=1 /proc/cmdline 
BOOT_IMAGE=/boot/vmlinuz-5.10.0-8-amd64 ro quiet ipv6.disable=1

$ ./builddir/ping/ping -c1 ::1
./builddir/ping/ping: ::1: Address family for hostname not supported

$ ./builddir/ping/ping -c1 127.0.0.1
PING 127.0.0.1 (127.0.0.1) 56(84) bytes of data.
64 bytes from 127.0.0.1: icmp_seq=1 ttl=64 time=0.088 ms

--- 127.0.0.1 ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 0.088/0.088/0.088/0.000 ms
```

> I think a better test, since we really only care about ::1 being available, is to check /proc/net/ipv6_route for a local route to that destination. An IPv6 localhost route will appear as a line starting with 00000000000000000000000000000001 in that file.

@nmeyerhans My plan was to check `/sys/module/ipv6/parameters/disable`. Using `/proc/net/ipv6_route` for a detection is a good idea, but `/proc/net/if_inet6` has probably easier format to read (`lo` is in `if_inet6` only once, format is also easier).
https://tldp.org/HOWTO/Linux+IPv6-HOWTO/ch11s04.html

And for longer term I plan to port LTP shell test library for tests and move testing code there (https://github.com/iputils/iputils/issues/243), which prevent me from introducing meson related bugs :).

**Comment by eli-schwartz** [comment]
> BTW while open to improvements I still don't see why this is a hard dependency: even with removed `ip` it's possible to build (tested on Debian stable):

I have checked as far back as meson 0.40 and current versions of meson error out when doing an upfront search for the program with something like:

```
Program or command 'ip' not found or not executable
```
while 0.40 errored out in Popen after catching FileNotFoundError and explicitly raising a mesonbuild.interpreterbase.InterpreterException (but allowing any returncode):

```
Could not execute command "ip".
```



> ```
> # mv /usr/bin/ip /usr/bin/ip_
> # ip
> -bash: /usr/bin/ip: No such file or directory`
> ```

This really tells me nothing. You could have a different version of `ip` elsewhere on your PATH, in theory, but the bash shell will not look for another version elsewhere because it has hashed the one at /usr/bin/ip for quick lookup, and bash violates the POSIX requirement to to search PATH again if the hashed version (a performance optimization *only*) is not found.

Bash when invoked in POSIX mode will conform to POSIX, and it will either successfully invoke another copy of the program in another directory, or it will report a different error:

```
bash: ip: command not found
```

`hash -d ip` will delete the "performance optimization" hash and cause bash to do a brand new lookup.

When attempting to see whether a command is available in PATH, none of this is irrelevant: the correct way to check if a program exists is NOT to try running it. Instead, use `type -a ip` and it will report every possible version of it anywhere on your PATH, or in shell aliases, or in shell functions, and will even report "ip: not found" instead of "/usr/bin/ip: no such file or directory" if no version of it exists anywhere else.

...

Can you confirm the `ip` command actually does not exist in your test, rather than bash tricking you? The output of builddir/meson-logs/meson-log.txt would be very interesting to see, since it actually logs the internals of run_command().

**Comment by pevik** [comment]
> This really tells me nothing. You could have a different version of `ip` elsewhere on your PATH, in theory, but the bash shell will not look for another version elsewhere because it has hashed the one at /usr/bin/ip for quick lookup, and bash violates the POSIX requirement to to search PATH again if the hashed version (a performance optimization _only_) is not found.

Sure, you don't have to believe I can reliably remove binary :). But really, there is no other `ip` on that system:

```
$ command -v ip; echo $?
1
```

Sure, running the default it fails as expected:
```
# rm -rf builddir/; meson builddir && make; echo $?
...
test/ping/meson.build:8:0: ERROR: Program or command 'ip' not found or not executable
...
1
```

But with it works on `-DSKIP_TESTS=true`:
```
# rm -rf builddir/; meson builddir -DSKIP_TESTS=true && make; echo $?
...
[19/19] Generating tracepath.8 with a custom command
...
0
```
At least it's working on 0.56.2. Maybe older versions have problem (when requiring meson 0.39, there was a warning: `WARNING: Project targeting '>=0.39' but tried to use feature introduced in '0.47.0': check arg in run_command.`, but that warning is not there when requiring version 0.40 - probably meson bug).

Anyway, I'm not arguing with you that it can be improved (either you send a PR or I'll fix it once get time). My whole exercise was to present workaround for code on current master.

**Comment by eli-schwartz** [comment]
> Sure, you don't have to believe I can reliably remove binary :). But really, there is no other `ip` on that system:

That is indeed a much better demonstration. :)

And, you probably do not want to know how absolutely often I've seen this sort of mistake, but there is a reason I prefer to be precise about using `type -a` and don't assume *anyone* is safe from making the mistake. It is... far too common, across a wide variety of experience levels.


> Maybe older versions have problem (when requiring meson 0.39, there was a warning `WARNING: Project targeting '>=0.39' but tried to use feature introduced in '0.47.0': check arg in run_command`., but that warning is not there when requiring version 0.40 - probably meson bug).

(For the record, my checking did not involve downloading iputils and trying to build it. I just traced the commit diffs checking to see how meson's run_command_impl interpreter method worked. I make no statements about the minimum compatible version this project builds using.)

...

Hmm, re-reading the original reporter's error message it's happening on line 38 of `ping/meson.build`, and I also overlooked that you're not saying meson ignores the missing binary -- you're saying that -DSKIP_TESTS=true doesn't even run that run_command in the first place.

But... commit 3163c49f9f4ad473a00d8a345ee334a028376011 both implemented the option you're suggesting in the first place, *and* moved the run_command from line 38 of `ping/meson.build` to line 7 of `ping/test/meson.build`. Later it got moved to `test/ping/meson.build`.

So @arekm

> Doesn't work. Still wants ip binary. Easy test by doing "mv ip ip.org".

It sounds like you tried passing an invalid option to an old version of iputils. You probably also used a version of meson that is older than meson 0.60 -- before that, meson would print a yellow warning that there is no "SKIP_TESTS" option, but meson 0.60 made that into a fatal error, specifically because there is never any good reason to try to specify a project option that isn't valid, and it's better to make that mistake immediately noticeable.

The workaround suggested by @pevik should definitely work if you upgrade iputils to version 20211215, but you seem to be using 20210722.
